pkg/virt-api/rest/objectgraph_test.go

```diff
@@ -0,0 +1,620 @@
+			graph = NewObjectGraph(virtClient, &v1.ObjectGraphOptions{})
+			node := graph.newGraphNode("test", "default", "invalid-resource", nil, false)
+			Expect(node).To(BeNil())
+		})
+	})
+
+	Context("with options", func() {
+		BeforeEach(func() {
+			vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+				Name: "test-instancetype",
+				Kind: "VirtualMachineInstancetype",
+			}
+			vm.Spec.Preference = &v1.PreferenceMatcher{
+				Name: "test-preference",
+				Kind: "VirtualMachinePreference",
+			}
+			vm.Spec.Template.Spec.AccessCredentials = []v1.AccessCredential{
+				{
+					SSHPublicKey: &v1.SSHPublicKeyAccessCredential{
+						Source: v1.SSHPublicKeyAccessCredentialSource{
+							Secret: &v1.AccessCredentialSecretSource{
+								SecretName: "test-secret",
+							},
+						},
+					},
+				},
+			}
+		})
+
+		It("should exclude optional resources when IncludeOptionalNodes is false", func() {
+			options := &v1.ObjectGraphOptions{
+				IncludeOptionalNodes: pointer.P(false),
+			}
+			graph = NewObjectGraph(virtClient, options)
+			graphNodes, err := graph.GetObjectGraph(vm)
+			Expect(err).NotTo(HaveOccurred())
+
+			childNames := make(map[string]bool)
+			for _, child := range graphNodes.Children {
+				childNames[child.ObjectReference.Name] = true
+			}
+
+			Expect(childNames).NotTo(HaveKey("test-instancetype"))
+			Expect(childNames).NotTo(HaveKey("test-preference"))
+			Expect(childNames).To(HaveKey("test-secret"))
+		})
+
+		It("should filter by label selector for config dependencies", func() {
+			options := &v1.ObjectGraphOptions{
+				LabelSelector: &k8smetav1.LabelSelector{
+					MatchLabels: map[string]string{
+						ObjectGraphDependencyLabel: "config",
+					},
+				},
+			}
+			graph = NewObjectGraph(virtClient, options)
+			graphNodes, err := graph.GetObjectGraph(vm)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Should only return config-related nodes
+			for _, child := range graphNodes.Children {
+				Expect(child.Labels[ObjectGraphDependencyLabel]).To(Equal("config"))
+			}
+		})
+
+		It("should filter by label selector for storage dependencies", func() {
+			vm.Spec.Template.Spec.Volumes = []v1.Volume{
+				{
+					Name: "test-volume",
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							PersistentVolumeClaimVolumeSource: corev1.PersistentVolumeClaimVolumeSource{
+								ClaimName: "test-pvc",
+							},
+						},
+					},
+				},
+			}
+
+			options := &v1.ObjectGraphOptions{
+				LabelSelector: &k8smetav1.LabelSelector{
+					MatchLabels: map[string]string{
+						ObjectGraphDependencyLabel: "storage",
+					},
+				},
+			}
+			graph = NewObjectGraph(virtClient, options)
+			graphNodes, err := graph.GetObjectGraph(vm)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Should only return storage-related nodes
+			for _, child := range graphNodes.Children {
+				Expect(child.Labels[ObjectGraphDependencyLabel]).To(Equal("storage"))
+			}
+		})
```

Comment:
Can you add a test to show that a recursive relationship is handled? Don't want an attacker to be able to DOS the apiserver